coderd/autobuild/executor/lifecycle_executor.go

```diff
@@ -89,77 +90,104 @@ func (e *Executor) runOnce(t time.Time) Stats {
-		eligibleWorkspaces, err := db.GetWorkspacesAutostart(e.ctx)
-		if err != nil {
-			return xerrors.Errorf("get eligible workspaces for autostart or autostop: %w", err)
-		}
 
-		for _, ws := range eligibleWorkspaces {
-			// Determine the workspace state based on its latest build.
-			priorHistory, err := db.GetLatestWorkspaceBuildByWorkspaceID(e.ctx, ws.ID)
-			if err != nil {
-				e.log.Warn(e.ctx, "get latest workspace build",
-					slog.F("workspace_id", ws.ID),
-					slog.Error(err),
-				)
-				continue
-			}
+	// TTL is set at the workspace level, and deadline at the workspace build level.
+	// When a workspace build is created, its deadline initially starts at zero.
+	// When provisionerd successfully completes a provision job, the deadline is
+	// set to now + TTL if the associated workspace has a TTL set. This deadline
+	// is what we compare against when performing autostop operations, rounded down
+	// to the minute.
+	//
+	// NOTE: If a workspace build is created with a given TTL and then the user either
+	//       changes or unsets the TTL, the deadline for the workspace build will not
+	//       have changed. This behavior is as expected per #2229.
+	eligibleWorkspaces, err := e.db.GetWorkspacesAutostart(e.ctx)
+	if err != nil {
+		e.log.Error(e.ctx, "get eligible workspaces for autostart or autostop", slog.Error(err))
+		return stats
+	}
 
-			priorJob, err := db.GetProvisionerJobByID(e.ctx, priorHistory.JobID)
-			if err != nil {
-				e.log.Warn(e.ctx, "get last provisioner job for workspace %q: %w",
-					slog.F("workspace_id", ws.ID),
-					slog.Error(err),
-				)
-				continue
-			}
+	// We only use errgroup here for convenience of API, not for early
+	// cancellation. This means we only return nil errors in th eg.Go.
+	eg := errgroup.Group{}
+	// Limit the concurrency to avoid overloading the database.
+	eg.SetLimit(10)
 
-			validTransition, nextTransition, err := getNextTransition(ws, priorHistory, priorJob)
-			if err != nil {
-				e.log.Debug(e.ctx, "skipping workspace",
-					slog.Error(err),
-					slog.F("workspace_id", ws.ID),
-				)
-				continue
-			}
+	for _, ws := range eligibleWorkspaces {
+		ws := ws
+		log := e.log.With(slog.F("workspace_id", ws.ID))
 
-			if currentTick.Before(nextTransition) {
-				e.log.Debug(e.ctx, "skipping workspace: too early",
-					slog.F("workspace_id", ws.ID),
-					slog.F("next_transition_at", nextTransition),
-					slog.F("transition", validTransition),
-					slog.F("current_tick", currentTick),
-				)
-				continue
-			}
+		eg.Go(func() error {
+			err := e.db.InTx(func(db database.Store) error {
+				var err error
 
-			e.log.Info(e.ctx, "scheduling workspace transition",
-				slog.F("workspace_id", ws.ID),
-				slog.F("transition", validTransition),
-			)
+				// Re-check eligibility since the first check was outside the
+				// transaction and the workspace settings may have changed.
+				ws, err = db.GetWorkspaceAutostart(e.ctx, ws.ID)
```

Comment:
Suggestion: we could just use `GetWorkspaceByID` and do the checks in Go, one less DB query to maintain.